### DISCUSSION TITLE
Slow first run for OneDrive with many files

### DISCUSSION THREAD

**Discussion by Ohaa** [abraunegg/onedrive#1891] [Q&A] [answered]
Hello,

I am on a dual-boot system (trying to switch to Linux permanently) and I am trying to set up OneDrive v.2.4.16 with my personal account which contains many small and some larger files (115k files, total of approx. 135 GB).

Syncing seems to work fine but is also very slow (2s per file). Two seconds per file would not be a problem when syncing 1-10 files during normal use but is a problem for the first synchronization when all files have to synced at once.

Since I am dual-booting, I tried copying the files directly from the other hard drive into the OneDrive folder, so that downloading can be skipped, but checking the files and folders seems to be an equally slow task (`Processing 263 OneDrive items to ensure consistent local state` repeating over and over).

Is there a way to speed up the first synchronization on a system with many small files?

Thanks in advance for your help and thanks a lot for your great work!

**Comment by abraunegg** [comment, marked as answer]
@Ohaa 
> Syncing seems to work fine but is also very slow (2s per file).

This is 100% normal due to 'curl' and the OneDrive API

> Since I am dual-booting, I tried copying the files directly from the other hard drive into the OneDrive folder, so that downloading can be skipped, but checking the files and folders seems to be an equally slow task (`Processing 263 OneDrive items to ensure consistent local state` repeating over and over).

This is 100% normal but, if the files already exist the application should be going through all the item batches very quickly. It will look like it is repeating to you - but in fact, it is processing X number of items as provided by OneDrive API in ~200 item lots (no way to change this - API limitation) to go through your entire 135K items that you have.

> Is there a way to speed up the first synchronization on a system with many small files?

* Copy all the files locally
* Let the application perform it's sync. If *all* the data already exists, it should not take long at all to process and validate 135K worth of items.

**Comment by ParamjitKahai** [comment]
The slowness is a major problem every time I sync files. I use a sync list to sync a limited number of files and folders. How is Microsoft able to sync that fast in Windows? Just a thought here, not a complaint or criticism.

**Reply by abraunegg** [reply]
@ParamjitKahai 
The issue is compounded for you by your use of 'sync_list'

Because you are using that as a method to include|exclude items to sync, a full traversal of OneDrive items is needed - because, OneDrive only sends changes when using their /delta query once ... and never again .... thus, if you add / remove / change data online - that change will be missed when using /delta

This is *why* you have this perception.

Additionally - ensure you are using v2.4.17 or later to ensure you have the most optimised sync operation experience.

**Reply by ParamjitKahai** [reply]
Yes, thank you! 'Sync_list' was the issue. And, yes, I am using v2.4.17.
